File: apps/preview/src/lib/vor/canonical.ts
/**
 * Deterministic canonical JSON — exact TS mirror of vor-license canonical.rs.
 *
 * Rules: object keys sorted lexicographically (byte order), no whitespace,
 * shortest RFC-8259 string escapes, integers stay integers.
 * The Ed25519 signature MUST cover exactly this byte form.
 */

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [k: string]: JsonValue };

/** Shortest valid JSON string escape (mirrors serde_json). */
function escapeJsonString(s: string): string {
  return JSON.stringify(s);
}

export function canonicalJson(value: JsonValue): string {
  let out = '';
  writeCanonical(value);
  return out;

  function writeCanonical(v: JsonValue): void {
    if (v === null) {
      out += 'null';
      return;
    }
    switch (typeof v) {
      case 'boolean':
        out += v ? 'true' : 'false';
        return;
      case 'number': {
        if (!Number.isFinite(v)) throw new Error('canonical: non-finite number');
        // serde_json round-trip form: integers without decimal point.
        if (Number.isInteger(v) && Math.abs(v) < 9.007199254740991e15) {
          out += String(v);
        } else {
          out += String(v);
        }
        return;
      }
      case 'string':
        out += escapeJsonString(v);
        return;
    }
    if (Array.isArray(v)) {
      out += '[';
      for (let i = 0; i < v.length; i++) {
        if (i > 0) out += ',';
        writeCanonical(v[i]);
      }
      out += ']';
      return;
    }
    // Object: deterministic ordering via byte-wise key sort.
    const keys = Object.keys(v).sort((a, b) => {
      const ea = new TextEncoder().encode(a);
      const eb = new TextEncoder().encode(b);
      const n = Math.min(ea.length, eb.length);
      for (let i = 0; i < n; i++) {
        if (ea[i] !== eb[i]) return ea[i] - eb[i];
      }
      return ea.length - eb.length;
    });
    out += '{';
    for (let i = 0; i < keys.length; i++) {
      if (i > 0) out += ',';
      out += escapeJsonString(keys[i]);
      out += ':';
      writeCanonical(v[keys[i]]);
    }
    out += '}';
  }
}

/** Parse then canonicalize in one step (mirrors canonicalize_str). */
export function canonicalizeStr(json: string): string {
  return canonicalJson(JSON.parse(json) as JsonValue);
}

File: core/vor-license/src/canonical.rs
use serde_json::Value;

/// Deterministic canonical JSON serialization.
///
/// Rules:
/// * Object keys sorted lexicographically (byte order).
/// * No whitespace between tokens.
/// * Strings escaped per RFC 8259 with the shortest valid form produced by
///   `serde_json`. Numbers keep their `serde_json` round-trip form.
///
/// The Ed25519 signature MUST cover exactly this canonical form so that any
/// byte-level modification of signed fields invalidates the license.
pub fn canonical_json(value: &Value) -> String {
    let mut out = String::new();
    write_canonical(value, &mut out);
    out
}

fn write_canonical(value: &Value, out: &mut String) {
    match value {
        Value::Null => out.push_str("null"),
        Value::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => out.push_str(&n.to_string()),
        Value::String(s) => {
            let esc = serde_json::to_string(s).unwrap_or_else(|_| "\"\"".into());
            out.push_str(&esc);
        }
        Value::Array(items) => {
            out.push('[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_canonical(item, out);
            }
            out.push(']');
        }
        Value::Object(map) => {
            // Deterministic ordering via BTreeMap, independent of serde_json features.
            let mut sorted: Vec<(&String, &Value)> = map.iter().collect();
            sorted.sort_by(|a, b| a.0.as_bytes().cmp(b.0.as_bytes()));
            out.push('{');
            for (i, (k, v)) in sorted.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                let esc = serde_json::to_string(k).unwrap_or_else(|_| "\"\"".into());
                out.push_str(&esc);
                out.push(':');
                write_canonical(v, out);
            }
            out.push('}');
        }
    }
}

/// Parse then canonicalize in one step.
pub fn canonicalize_str(json: &str) -> Result<String, serde_json::Error> {
    let v: Value = serde_json::from_str(json)?;
    Ok(canonical_json(&v))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn sorts_keys_and_strips_whitespace() {
        let v = json!({ "b": 1, "a": { "z": true, "y": [3, 1] } });
        assert_eq!(canonical_json(&v), r#"{"a":{"y":[3,1],"z":true},"b":1}"#);
    }

    #[test]
    fn is_stable_across_input_ordering() {
        let a = canonicalize_str(r#"{"x":1,"y":"س"}"#).unwrap();
        let b = canonicalize_str(r#" { "y" : "س" , "x" : 1 } "#).unwrap();
        assert_eq!(a, b);
        assert_eq!(a, r#"{"x":1,"y":"س"}"#);
    }
}
